infrastructure/edgeRedirectLambdaAssociations.ts

```diff
@@ -0,0 +1,124 @@
+import * as aws from "@pulumi/aws";
+import * as URLPattern from "url-pattern";
+import { CloudFrontRequestEvent } from "aws-lambda";
+import { LambdaEdge } from "./lambdaEdge";
+
+export function getEdgeRedirectAssociations():
+    aws.types.input.cloudfront.DistributionDefaultCacheBehaviorLambdaFunctionAssociation[] {
+    // https://aws.amazon.com/blogs/networking-and-content-delivery/handling-redirectsedge-part1/
+
+    const lambdaEdge = new LambdaEdge(
+        "redirects", {
+            func: (event: CloudFrontRequestEvent, context, callback) => {
+
+                // Get contents of response.
+                const request = event.Records[0].cf.request;
+
+                const redirect = getRedirect(request.uri);
+                if (redirect) {
+                    const modifiedResponse = {
+                        status: "301",
+                        statusDescription: 'Moved Permanently',
+                        headers: {
+                            "location": [
+                                {
+                                    key: "Location",
+                                    value: redirect,
+                                },
+                            ],
+                            "cache-control": [
+                                {
+                                    key: "Cache-Control",
+                                    value: "max-age=3600",
+                                },
+                            ],
+                        },
+                    };
+
+                    // Use a modified response.
+                    callback(null, modifiedResponse);
+
+                } else {
+
+                    // Use the original request.
+                    callback(null, request)
+                }
+            },
+            funcDescription: "Lambda function that conditionally redirects based on a path-matching expression.",
+        },
+        {}
+    );
+
+    return [
+        {
+            includeBody: false,
+            lambdaArn: lambdaEdge.getLambdaEdgeArn(),
+            eventType: "origin-request",
+        }
+    ];
+}
+
+function getRedirect(uri: string): string | undefined {
+    return nodeSDKRedirect(uri) || pythonSDKRedirect(uri) || dotnetSDKRedirect(uri) || undefined;
+}
+
+function nodeSDKRedirect(uri: string): string | undefined {
+    const pattern = new URLPattern("/docs/reference/pkg/nodejs/pulumi/:provider(/:service)(*)");
```

Comment:
Could these `URLPattern` instances be declared as `const`s outside the function since they don't seem to be dynamic per request? We could avoid having to instantiate them for every request.

Reply:
Ah, I see what you mean. They could, but I think I'd prefer to keep them close to the code that refers to them, just for readability. (It just translates the string into a regex; the execution time looks like somewhere between 0 and 1 ms.)

Reply:
> t just translates the string into a regex

I would definitely put them outside the function since this will run for every single request the website receives in order to check if the redirect needs to happen. You could put literally outside each respective function and still maintain readability I think. But then again it's a Lambda and I am not sure how AWS will reuse function contexts in this scenario. Maybe it's ok, maybe it's not 🤷‍♂️ I would err on the side of putting it outside since it doesn't affect readability all that much anyway.

Reply:
Ah I just noticed that you are using an `origin-request` (and not `viewer-request`) trigger so this function wouldn't run for every request that CFN receives, since CloudFront will [cache the redirect response](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/http-3xx-status-codes.html) for the requested object based on the `cache-control` header. So feel free to leave it as-is. I would recommend bumping that `max-age`, though. 1 hour seems low for pages that have moved permanently, no?

Reply:
I hear ya. 🤔  I do find it slightly more readable the way it's written now -- but separate from that, to your point, since everything ends up serialized into the function body anyway, and the function is run on every request, then if the `URLPatterns` were defined outside of their functions, every request would instantiate all three of them, whereas as written now, if the URL were a `nodejs` URL, you'd instantiate the Node pattern but not the Python or .NET patterns, since `getRedirect` would short-circuit on the first match.

Then again, since most requests won't be for redirectible URLs anyway, most will end up instantiating all three -- so maybe the thing to do instead is just return early unless the URL matches `/docs/reference/pkg/` to begin with. Lemme try that. What do you think?

Reply:
This approach (of exiting early when the request isn't for an SDK doc) seems to work. I'll go with it for now -- thanks for prompting the additional thinking on it. Take a look and let me know if you have any objections.

On the cache-control bit, I honesty just used what was in the AWS example. But thinking about it a bit more, it looks like omitting the cache-control header altogether would make the browser cache the 301 ~permanently (i.e., for as long as it can), which sounds like what we want -- although with the caveat that that we wouldn't be able to republish at those URLs (since we wouldn't be able to invalidate the browser cache at that point).  Since that possibility seems low but not totally zero, I've opted to set this at a week, just to err on the side of having that smidge of extra control. Lemme know what you think.

Reply:
> Since that possibility seems low but not totally zero, I've opted to set this at a week, just to err on the side of having that smidge of extra control. Lemme know what you think.

👍 That's along the lines of what I was thinking too.